File: Gambrinus/ObjectModel/ObjectModel+Settings.m
/*
* Copyright 2015 Coodly LLC
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#import <JCSFoundation/NSDate+ISO8601.h>
#import "ObjectModel+Settings.h"
#import "Setting.h"
#import "Constants.h"

typedef NS_ENUM(short, SettingKey) {
    SettingLastVerifiedPullDate,
    SettingSortOrder
};

@implementation ObjectModel (Settings)

- (void)setLastVerifiedPullDate:(NSDate *)date {
    [self setSettingValue:[date iso8601String] forKey:SettingLastVerifiedPullDate];
}

- (NSDate *)lastVerifiedPullDate:(NSDate *)defaultDate {
    return [self dateSettingWithKey:SettingLastVerifiedPullDate defaultValue:defaultDate];
}

- (PostsSortOrder)postsSortOrder {
    return (PostsSortOrder) [self shortValueForKey:SettingSortOrder defaultValue:OrderByDateDesc];
}

- (void)setPostsSortOrder:(PostsSortOrder)order {
    [self setShortValue:order forKey:SettingSortOrder];
    [self saveContext:^{
        [[NSNotificationCenter defaultCenter] postNotificationName:GambrinusSortOrderChangedNotification object:nil];
    }];
}

- (short)shortValueForKey:(SettingKey)key defaultValue:(short)defaultValue {
    Setting *setting = [self loadSettingWithKey:key];
    if (!setting) {
        return defaultValue;
    }
    return (short) [setting.value intValue];
}

- (void)setShortValue:(short)value forKey:(SettingKey)key {
    Setting *setting = [self loadSettingWithKey:key];
    if (!setting) {
        setting = [Setting insertInManagedObjectContext:self.managedObjectContext];
        [setting setKeyValue:key];
    }

    [setting setValue:[NSString stringWithFormat:@"%d", value]];
}

- (NSDate *)dateSettingWithKey:(SettingKey)key defaultValue:(NSDate *)defaultValue {
    Setting *setting = [self loadSettingWithKey:key];
    if (!setting) {
        return defaultValue;
    }

    NSDate *result = setting.dateValue;
    if (!result) {
        result = defaultValue;
    }

    return result;
}

- (void)setSettingValue:(NSString *)value forKey:(SettingKey)key {
    Setting *setting = [self loadSettingWithKey:key];
    if (!setting) {
        setting = [Setting insertInManagedObjectContext:self.managedObjectContext];
        [setting setKeyValue:key];
    }

    [setting setValue:value];
}

- (Setting *)loadSettingWithKey:(SettingKey)key {
    NSPredicate *keyPredicate = [NSPredicate predicateWithFormat:@"key = %d", key];
    return [self fetchEntityNamed:[Setting entityName] withPredicate:keyPredicate];
}

@end

File: Gambrinus/ObjectModel/ObjectModel+Settings.swift
/*
 * Copyright 2015 Coodly LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

enum Key: Int {
    case lastVerifiedPullDate
    case sortOrder
}

enum PostsSortOrder: Int {
    case byDateDesc
    case byDateAsc
    case byPostName
    case byRBBeerName
    case byRBScore
    case byStyle
    case byBrewer
    
    func reversed() -> PostsSortOrder? {
        switch self {
        case .byDateAsc:
            return .byDateDesc
        case .byDateDesc:
            return .byDateAsc
        default:
            return nil
        }
    }
}

extension ObjectModel {
    func sortOrder() -> PostsSortOrder {
        if let setting = setting(for: .sortOrder), let code = Int(setting.value), let value = PostsSortOrder(rawValue: code) {
            return value
        }
        
        return .byDateDesc
    }
    
    func setSortOrder(_ value: PostsSortOrder) {
        let saved: Setting = setting(for: .sortOrder) ?? managedObjectContext.insertEntity()
        saved.key = NSNumber(integerLiteral: Key.sortOrder.rawValue)
        saved.value = "\(value.rawValue)"

        saveContext() {
            NotificationCenter.default.post(name: Notification.Name("GambrinusSortOrderChangedNotification"), object: nil)
        }
    }
    
    func lastKnownPullDate() -> Date {
        return date(for: .lastVerifiedPullDate)
    }
    
    func setPostsRefreshTime(_ date: Date) {
        save(date: date, for: .lastVerifiedPullDate)
    }
    
    private func save(date: Date, for key: Key) {
        let saved = setting(for: key) ?? managedObjectContext.insertEntity()
        saved.key = NSNumber(integerLiteral: key.rawValue)
        saved.value = (date as NSDate).iso8601String()
    }
    
    private func date(for key: Key, defaultValue: Date = Date.distantPast) -> Date {
        if let existing = setting(for: key) {
            return existing.dateValue()
        }
        
        return defaultValue
    }
    
    private func setting(for key: Key) -> Setting? {
        return managedObjectContext.fetchEntity(where: "key", hasValue: key.rawValue as AnyObject)
    }
}
